EIPS/eip-draft-whisper.md

```diff
@@ -0,0 +1,138 @@
+
+**P2P Request** [`0x03`, `whisper_envelope`]
+
+This packet contains two objects: integer (0x03) followed by a single Whisper Envelope.
+
+This packet is used for sending Dapp-level peer-to-peer requests, e.g. Whisper Mail Client requesting old messages from the Whisper Mail Server.
+
+**PoW Requirement** [`0x04`, `PoW`]
+
+This packet contains two objects: integer (0x04) followed by a single floating point value of PoW.
+
+This packet is used by Whisper nodes for dynamic adjustment of their individual PoW requirements. Receipient of this message should no longer deliver the sender messages with PoW lower than specified in this message.
+
+**Bloom Filter** [`0x05`, `bytes`]
+
+This packet contains two objects: integer (0x05) followed by a byte array of arbitrary size.
+
+This packet is used by Whisper nodes for sharing their interest in messages with specific topics.
+
+### Whisper Envelope
+
+Envelopes are RLP-encoded structures of the following format:
+
+	[ Version, Expiry, TTL, Topic, AESNonce, Data, EnvNonce ]
+	
+`Version`: up to 4 bytes (currently one byte containing zero). Version indicates encryption method. If Version is higher than current, envelope could not be decoded, and therefore only forwarded to the peers.
+
+`Expiry`: 4 bytes (UNIX time in seconds).
+
+`TTL`: 4 bytes (time-to-live in seconds).
+
+`Topic`: 4 bytes of arbitrary data.
+
+`AESNonce`: 12 bytes of random data (only present in case of symmetric encryption).
+
+`Data`: byte array of arbitrary size (contains encrypted message).
+
+`EnvNonce`: 8 bytes of arbitrary data (used for PoW calculation).
+
+### Contents of Data Field (Message)
+
+Data field contains encrypted message of the Envelope. Plaintext (unencrypted) payload is formed as a concatenation of a single byte for flags, additional metadata (as stipulated by the flags) and the actual payload. The message has the following structure:
+
+    flags: 1 byte
+    
+    optional padding: byte array of arbitrary size
+    
+    payload: byte array of arbitrary size
+    
+    optional signature: 65 bytes
+
+Those unable to decrypt the message data are also unable to access the signature. The signature, if provided, is the ECDSA signature of the Keccak-256 hash of the unencrypted data using the secret key of the originator identity. The signature is serialised as the concatenation of the `R`, `S` and `V` parameters of the SECP-256k1 ECDSA signature, in that order. `R` and `S` are both big-endian encoded, fixed-width 256-bit unsigned. `V` is an 8-bit big-endian encoded, non-normalised and should be either 27 or 28. 
+
+The padding is introduced in order to align the message size, since message size alone might reveal important metainformation. The first several bytes of padding (up to four bytes) indicate the total size of padding (including these length bytes). E.g. if padding is less than 256 bytes, then one byte is enough; if padding is less than 65536 bytes, then 2 bytes; and so on.
```

Comment:
if we leave it to the Dapp level, very few people will implement it, and those who do will look very suspicious. this is the whole point of plausible deniability -- it should be implemented by default and used by everyone.